src/edgepi/peripherals/gpio.py

```diff
@@ -45,6 +45,57 @@ def read_state(self):
         """
         return self.gpio.read()
 
+    def open_read_state(self, pin_num:int, pin_dir:str, pin_bias:str):
+        '''
+        To minimize issues with the lock, we open & read in a single function call
+        '''
+        try:
+            GpioDevice.lock_gpio.acquire()
+            gpio   = GPIO(self.gpio_fd, pin_num, pin_dir, bias=pin_bias)
+            result = gpio.read()
+
+        finally:
+            try:
+                gpio.close()
+            except Exception as exc:
+                raise OSError(f"Failed to close {self.gpio_fd}") from exc
+            finally:
+                GpioDevice.lock_gpio.release()
+
+        return result
+
+    def open_read_state_batch(self, pin_num_list:list[int], pin_dir:str, pin_bias:str) -> list:
+        '''
+        Batch several gpio reads into a single lock / unlock. We can also take advantage of the fact we're accessing
+        the gpio only from different pins, and so we can open & close just a single time.
+        '''
+        results = []
+        try:
+            GpioDevice.lock_gpio.acquire()
+            gpio = None
+            try:
+                for pin_num in pin_num_list:
+                    if gpio is None:
+                        gpio = GPIO(self.gpio_fd, pin_num, pin_dir, bias=pin_bias)
+                    else:
+                        # NOTE: we'll need to be careful when we update periphery, since we depend on private functionality
+                        # pylint: disable=protected-access
+                        gpio._line = pin_num
+                        # pylint: disable=protected-access
+                        gpio._reopen(pin_dir, edge="none", bias=pin_bias, drive="default", inverted=False)
```

Comment:
this change enables the batching of gpio reads, which minimizes the number of device open calls (I/O), which improves performance quite a bit. One tradeoff is that we touch private variables & functions, which could change if the library is updated. It seems the file these variables are in hasn't been modified for a year, but still something to keep in mind.